pkg/reconciler/route/resources/alg_types.go

```diff
@@ -0,0 +1,112 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resources
+
+import (
+	"github.com/google/kf/pkg/kf/algorithms"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"knative.dev/pkg/apis/istio/v1alpha3"
+)
+
+// TODO(poy): This file SHOULD be generated, but it has been written by hand.
```

Comment:
I think these algorithms would be much more powerful if they were fully generated for the specific type to preserve type-safety.

Reply:
Interesting, so instead of generating the implementing interface, generate the full algorithm?

Reply:
Yeah, let's get this in for now and we can discuss later. It might be worth creating an independent project for generators to do that. We already have some infrastructure for doing functional things to the resources.